<html>
<head>
	<title>README</title>

<style type="text/css">
	#rt{text-align:right; color: #000000; font-weight: bold}
	#grBk {background-color: #CC9966;}
	TD {font-family: Verdana, Arial, Helvetica, sans-serif; font-size:12;}
	H1 {color: #996633; font:arial; font-size:16;}
	H2 {color: #996633; font:arial; font-size:12;}
	BODY {color:black; background-color:white; font-family: Verdana, Arial, Helvetica, sans-serif; font-size:12;}
	
	A:link    {text-decoration: none; color: #000000; font-weight: bold}  
	A:visited {text-decoration: none; color: #000000; font-weight: bold}   
	A:hover   {text-decoration: none; color: #FFCC66; font-weight: bold} 
	A:active  {text-decoration: none; color: #000000; font-weight: bold}   
</style>
</head>
<center><H1>USeq</H1></center>
<p>
<h2>Disclaimer:</h2> This software is distributed with no warrantee or guarantee of fitness. See the open source <A HREF="license.txt">BSD License</A></b>.<p>

<h2>What is USeq?</h2>
USeq is a collection of software tools for for both low and high level analysis of next generation, ultra high throughput signature sequencing data from the Solexa, SOLiD, and 454 platforms. 
Initial emphasis: chIP-seq and RNA-Seq with FDR estimations. USeq is under continuous development at the <a href="http://www.hci.utah.edu/">
Huntsman Cancer Institute</a> in the <a href="http://bioserver.hci.utah.edu/BioInfo/">Utah Bioinformatics Shared Resource Center</a>. 
<p>
A description of the statistics and comparison of methods used in USeq is now published, see <A HREF="http://www.biomedcentral.com/1471-2105/9/523">Nix DA, Courdy SJ, Boucher KM: Empirical methods for controlling false positives and estimating confidence in chIP-seq peaks. BMC Bioinformatics. 2008 Dec 5;9(1):523.<a>
Note, USeq now makes use of the <a href="http://www-huber.embl.de/users/anders/DESeq/">DESeq R package</a> to calculate p-values from a negative binomial distribution for differential enrichment/ reduction. See <A HREF="http://dx.doi.org/10.1038/npre.2010.4282.1"> Anders S, Huber W: 
Differential expression analysis for sequence count data. Nature Precedings. 2010 Jan </a>  
<p>
Read about USeq's 1st and 2nd place awards in the ChIP-Seq Community Challenge! <A HREF="nextGenJCMar09.ppt">PowerPoint Presentation </A>, <a href="http://seqanswers.com/forums/showthread.php?t=1039">SEQanswers Forum</a>
<h2>How to use/ install USeq?</h2>
To use USeq, download the <a href="http://sourceforge.net/projects/useq">latest release</a>, install <a href="http://www.java.com/en/download/">Java 1.6+</a>, <a href="http://www.r-project.org/">R</a>, Storey's <a href="http://genomics.princeton.edu/storeylab/qvalue/">Q-Value</a> and Anders' <a href="http://www-huber.embl.de/users/anders/DESeq/">DESeq</a> R libraries and launch the java jar applications with the -jar option.<br>
For example, to execute 'ScanSeqs' type 'java -jar yourPathTo/USeq/Apps/ScanSeqs' on the command line to print a menu and description of options. <p>
Afraid of the command line interface? No problem, give the GUI_ClickMe.jar a try. Launch it on Mac's and Windows by double clicking the application. This runs a draggable/ point and click wrapper 
around the USeq applications. <p>
Read the <A HREF="usage.html">Usage</A> guide. It contains detailed step-by-step instructions in how to use USeq applications to process your RNA-Seq and ChIP-Seq datasets.<p>
Join the <a href="https://lists.sourceforge.net/lists/listinfo/useq-users">USeq Users mailing list</a> to stay up to date with the latest USeq developments, tips, and tricks for next generation genomic analysis.<p>

<h2>Requirements</h2>
 <ul>
 	<li>MacOSX or Linux, USeq apps are untested on Windows
	<li><a href="http://www.java.com/en/download/">Java 1.6+</a>
	<li><a href="http://www.r-project.org/">R</a> with Storey's <a href="http://genomics.princeton.edu/storeylab/qvalue/">Q-Value</a> and Anders' <a href="http://www-huber.embl.de/users/anders/DESeq/">DESeq</a> libraries.
	<li>Familiarity with command line programing
</ul>	

<H2>Documentation:</H2>
 <ul>
	<li><A HREF="usage.html">Usage</A> - Contains detailed tutorials in how to use USeq to process your sequencing data (ChIP-Seq, <a href="usageRNASeq.html">RNA-Seq</a>, <a href="usageBisSeq.html">Bis-Seq</a> etc).</li> 
	<li><A HREF="applications.html">Applications</A> - A listing of the various USeq applications.</li>
	<li><A HREF="cmdLnMenus.html">Menus</A> - Command line menus/ instructions/ parameters for each of the applications.</li>
	<li><A HREF="releaseNotes.txt">Release Notes</A> - Changes associated with each release.</li>
	<li><A HREF="outputFileTypeDescriptions.html">Output File Type Descriptions</A> - Description of the different types of file output generated by USeq applications.</li>
	<li><A HREF="useqArchiveFormat.html">USeq Binary Archive Format</A> - Description of the USeq Binary Archive format for storing and distributing massive genomic datasets.</li>
	<li><A HREF="buildingUSeqFromEclipseSVN.txt">Developers Guide</A> - Guide for java developers to create a USeq Eclipse project linked to the Sourceforge SVN.</li>
	<li><A HREF="license.txt">BSD License</A> - The BSD open source USeq license.</li>
</ul>	
	
<H2>Misc:</H2>
 <ul>
	<li><a href="http://bioserver.hci.utah.edu/Annotations/">Annotation Files </a> - Useful gene annotation tables.</li>
	<li><a href="http://bioserver.hci.utah.edu/TestDatasets/">Test Datasets</a> - Test datasets for benchmarking Next Gen Seq algorithms.</li>
</ul>

<H2>Common mistakes:</H2>
 <ul>
	<li>No java sourced or installed.  Type 'java -version' on the command line.  It should be 1.6+ </li> 
	<li>Not using the -jar option.  Type: 'java -jar -Xmx2G yourPathTo/USeq/Apps/ScanSeqs'</li> 
	<li>Out of memory error.  Increase the heap size with the -Xmx flag.  Type: 'java -jar -Xmx3G yourPathTo/USeq/Apps/ScanSeqs'</li> 
	<li>Trying to run the apps on Windows.  It is worth a try. These have been tested on Linux and Mac OSX.</li>
</ul>

Questions? Comments? Post messages to the <a href="mailto:useq-users@lists.sourceforge.net">USeq Users Email List</a> or, if needed, contact <a href="mailto:David.Nix@hci.utah.edu">David Nix</a> directly in the <a href="mailto:bioinformaticsCore@utah.edu">Bioinformatics Core</a>. 
<p>
Many thanks to <a href="mailto:ken.boucher@hci.utah.edu">Ken Boucher<a/> in the <a href="http://www.hci.utah.edu/research/sharedResources.jsp?deptId=5">Biostatistics Shared Resource<a/> for help with the statistical methods.
<p>
Want to contribute? Join the <a href="https://lists.sourceforge.net/lists/listinfo/useq-users">USeq Users mailing list</a> to stay up to date with the latest USeq developments, tips, and tricks for next generation genomic analysis.<p>
<p>
<center><a href="http://sourceforge.net"><img src="http://sflogo.sourceforge.net/sflogo.php?group_id=222996&amp;type=2" width="125" height="37" border="0" alt="SourceForge.net Logo" /></a></center>

</HTML>
